### PR TITLE
THRESH Awaiting Review

### PR DESCRIPTION
**Pull request by FarzanT**
* No longer creates multiple edges
* TODO: Resolve multiple components

### CODE REVIEW

File: R/edgeThresh.R [hunk 1]
```diff
@@ -0,0 +1,461 @@
+# edgeThresh.R
+
+#' Define edge-removal threshold parameter δ for each target subnet-order, from
+#' random permutations of the input EGG object
+#'
+#' \code{THRESH} creates N permutations of the input graph EGG by randomly
+#' choosing 2 edges and swapping their target endpoints (i.e. B in the
+#' edge A → B). Finds the value δ (edge weight) for each random graph such that
+#' the maximum size of any component in the graph is Lmax, and returns the
```
Comment (hyginn) [inline R/edgeThresh.R]: This description is not correct.

File: R/edgeThresh.R [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+#' median of N δ values.
+#'
+#' @section Swapping Edges:
+#'     Bidirectional edges must only be swapped with each other, Since all 4
+#'     incident nodes in 2 bidirectional edges are receiving endpoints, one node
+#'     will be randomly selected from each edge, and a new edge is formed
+#'     between the selected as well as the remaining vertices.
+#'     At each unidirectional edge swap, only the receiving endpoints are
+#'     swapped e.g. (A → B, C → D) to (A → D, C → B).
+#'     In both cases, a swap should NOT result in:
```
Comment (hyginn) [inline R/edgeThresh.R]: You don't need to describe what you are **not** doing ...

File: R/edgeThresh.R [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+#'     If the current permutation's ord values is less than Lmin (without
+#'     finding δ), the delta is skipped, and interpolated later on.
+#'
+#' @param EGG An EGG object - the input heat-equilibrated gene score graph.
+#' @param Lmax	Maximum order for the largest connected component in a random
+#' network. See details.
+#' @param Lmin Minimum order for the largest connected component. Defaults to 2.
+#' @param N	The number of permuted networks to produce for choosing a δ
+#' cutoff.
+#' @param Q	The number of edge-swapping permutations to produce one of the
```
Comment (RicardoRam) [inline R/edgeThresh.R]: I think parameters N and Q should be renamed to be more descriptive.

File: R/edgeThresh.R [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+#'     another vertex (B) is switched with another incoming edge to vertex B,
+#'     resulting in (A) having two outgoing edges to (B).
+#'
+#'        ( )     ( )                  ( )  ←  ( )
+#'         ↑       ↓             →
+#'        (A)  →  (B)                  (A)   ⇉  (B)
+#'
+#'     "Self-edges": e.g. (A ↔ B, C ↔ A) to (A ↔ A, B ↔ C)
+#'     Where in the case of:
+#'     'Unidirectional edges': Precedent of one edge is the antecedent of the
```
Comment (hyginn) [inline R/edgeThresh.R]: Plain English please.

File: R/edgeThresh.R [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+#'     After creating the permutated graph, the following procedure is run on
+#'     the graph for each Lmax starting from Lmin to the Lmax arguments.
+#'     Initially, δ values are chosen by finding min(|E|, 100) quantiles of
+#'     edge weight distribution, and used to delete edges from the permutation.
+#'     As soon as the largest component has size (ord) between Lmin and Lmax,
+#'     If the current permutation's ord values is less than Lmin (without
+#'     finding δ), the delta is skipped, and interpolated later on.
+#'
+#' @param EGG An EGG object - the input heat-equilibrated gene score graph.
+#' @param Lmax	Maximum order for the largest connected component in a random
```
Comment (RicardoRam) [inline R/edgeThresh.R]: include default behavior for Lmax

File: R/edgeThresh.R [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+#'
+#'        ( )     ( )                  ( )  ←  ( )
+#'         ↑       ↓             →
+#'        (A)  →  (B)                  (A)   ⇉  (B)
+#'
+#'     "Self-edges": e.g. (A ↔ B, C ↔ A) to (A ↔ A, B ↔ C)
+#'     Where in the case of:
+#'     'Unidirectional edges': Precedent of one edge is the antecedent of the
+#'     other edge to be swapped.
+#'     'Bidirectional edges': The edge pair to be swapped involves less than 4
```
Comment (hyginn) [inline R/edgeThresh.R]: ??? That's not a valid swap.

File: R/edgeThresh.R [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+            currentGraph = igraph::delete.edges(currentGraph, edgesToRemove)
+
+            # Find the size of the largest strongly connected component, ord
+            ord <- max(igraph::components(currentGraph, "strong")$csize)
+
+            # If largest component has size less than Lmax, continue incrementing δ
+            # else, stop and record δ in the vector, and increase ord
+            # by 1
+            if (ord < Lmin) {
+                break
```
Comment (RicardoRam) [inline R/edgeThresh.R]: I think you mean next here (unless you wanted to exit the for loop when ord < Lmin)

File: R/edgeThresh.R [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+# (per graph, or for the whole process?)
+
+
+
+
+# ==== Graph Permutation =======================================================
+#' Helper function to create a valid permutation of the EGG graph.
+#' Only permutates the largest connected component
+#' @param EGG The given EGG graph from parent function.
+#' @param Q The number of edge-swapping permutations to produce the
```
Comment (RicardoRam) [inline R/edgeThresh.R]: Again, a more descriptive variable name might be better

File: R/edgeThresh.R [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+
+    return(mainComponent)
+
+}
+
+#' interpolates NA values in a vector by interpolateing from two
+#' bracketing non-missing values. If is.na(v[1]), the left-bracket is
+#' vMin, if is.na(v[length(v)]), vMax is the right bracket.
+#' @param v a numeric vector
+#' @param vMin numeric
```
Comment (RicardoRam) [inline R/edgeThresh.R]: what is vMin?

File: R/edgeThresh.R [hunk 10]
```diff
@@ -0,0 +1,461 @@
...
+#' @param vMax numeric
```
Comment (RicardoRam) [inline R/edgeThresh.R]: what is vMax?

File: tests/testthat/testEdgeThresh.R [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+nVertices <- igraph::vcount(EGG)
+nEdges <- igraph::ecount(EGG)
+# ==== END SETUP AND PREPARE ===================================================
+
+
+test_that("edgeThresh parameter errors are correctly handled", {
+    # Try each parameter missing in turn (only EGG has no defaults)
+    expect_error(edgeThresh())
+
+    # Try parameters out of expected bounds
```
Comment (RicardoRam) [inline tests/testthat/testEdgeThresh.R]: Include error messages with expect_error calls

File: R/edgeThresh.R [hunk 12]
```diff
@@ -0,0 +1,461 @@
...
+#'         ↑       ↓             →
+#'        (A)  →  (B)                  (A)   ⇉  (B)
+#'
+#'     "Self-edges": e.g. (A ↔ B, C ↔ A) to (A ↔ A, B ↔ C)
+#'     Where in the case of:
+#'     'Unidirectional edges': Precedent of one edge is the antecedent of the
+#'     other edge to be swapped.
+#'     'Bidirectional edges': The edge pair to be swapped involves less than 4
+#'     distinct nodes.
+#'     "Disconnected components": Considering (A → B, C → D), since A and B are
```
Comment (hyginn) [inline R/edgeThresh.R]: This is false! Consider the six - cycle X-A-B-Y-C-D-(X): swap (A-B), (C-D) -> (A-D), (C-B) . Therefore you now have the two three-cycles X-A-D-(X) and Y-C-B-(Y). We discussed this example!

File: R/edgeThresh.R [hunk 13]
```diff
@@ -0,0 +1,461 @@
...
+    # cR <-
+    #     c(cR, .checkArgs(writeLog,    like = logical(1), checkSize = TRUE))
+
+    # if (length(cR) > 0) {
+    #     stop(cR)
+    # }
+    # Lmax cannot be more than graph size
+    # Validate the Lmax parameter
+    if (!missing(Lmax) &&
+        (Lmax > (numVertices <- igraph::vcount(EGG)))) {
```
Comment (hyginn) [inline R/edgeThresh.R]: You are not following the algorithm I have outlined for you as pseudocode. I am stopping the review at this point. We need to talk.

File: tests/testthat/testEdgeThresh.R (hunk 11 above)
Comment (RicardoRam) [inline tests/testthat/testEdgeThresh.R]: The convention that I have seen so far has been to separate tests for each parameter into sections and test one thing at a time in each test. Can you add comments for what is being tested in each expect_error test?

**Comment by RicardoRam** [comment]
Please paste R CMD check output here

File: R/edgeThresh.R (hunk 3 above)
Comment (FarzanT) [inline R/edgeThresh.R]: @RicardoRam This is what the design enforces, can I change what is specified on the task page?

File: R/edgeThresh.R (hunk 6 above)
Comment (FarzanT) [inline R/edgeThresh.R]: @hyginn Please consider the context.
```
"Self-edges": e.g. (A ↔ B, C ↔ A) to (A ↔ A, B ↔ C)
#'     Where in the case of:
#'     'Unidirectional edges': Precedent of one edge is the antecedent of the
#'     other edge to be swapped.
#'     'Bidirectional edges': The edge pair to be swapped involves less than 4
#'     distinct nodes.
```

File: R/edgeThresh.R (hunk 12 above)
Comment (FarzanT) [inline R/edgeThresh.R]: @hyginn Again, please consider the context:
```
In both cases, a swap should NOT result in:
```
Also, I say the A and B __should__ remain in the same component. i.e. that is the goal. Now, what is wrong with the statement?

File: R/edgeThresh.R (hunk 7 above)
Comment (FarzanT) [inline R/edgeThresh.R]: @RicardoRam Break, since this graph has no more components of the size larger than Lmin.

File: R/edgeThresh.R (hunk 9 above)
Comment (FarzanT) [inline R/edgeThresh.R]: @RicardoRam This is a function made by Prof. Steipe for interpolating NA values, I simply copied it here.
It is the smallest value expected for the interpolation.

File: R/edgeThresh.R (hunk 10 above)
Comment (FarzanT) [inline R/edgeThresh.R]: @RicardoRam The largest value expected for the interpolation.
I'll add these descriptions to the function, but I was hoping that this function gets added to _utils_ later on, as it is not specific to THRESH.

File: tests/testthat/testEdgeThresh.R (hunk 11 above)
Comment (FarzanT) [inline tests/testthat/testEdgeThresh.R]:
```
Try each parameter missing in turn (only EGG has no defaults)
```
edgeThresh is called with no input parameters and thus must result in an error.

**Comment by FarzanT** [comment]
@RicardoRam R CMD check added 
```
==> devtools::check(document = FALSE)

Setting env vars ---------------------------------------------------------------
CFLAGS  : -Wall -pedantic
CXXFLAGS: -Wall -pedantic
Building rete ------------------------------------------------------------------
'/Library/Frameworks/R.framework/Resources/bin/R' --no-site-file --no-environ  \
  --no-save --no-restore --quiet CMD build '/Users/farzantaj/Second  \
  Semester/BCB420/Farzan-rete' --no-resave-data --no-manual 

* checking for file ‘/Users/farzantaj/Second Semester/BCB420/Farzan-rete/DESCRIPTION’ ... OK
* preparing ‘rete’:
* checking DESCRIPTION meta-information ... OK
* checking for LF line-endings in source and make files
* checking for empty or unneeded directories
* building ‘rete_0.1.0.tar.gz’

Setting env vars ---------------------------------------------------------------
_R_CHECK_CRAN_INCOMING_ : FALSE
_R_CHECK_FORCE_SUGGESTS_: FALSE
Checking rete ------------------------------------------------------------------
'/Library/Frameworks/R.framework/Resources/bin/R' --no-site-file --no-environ  \
  --no-save --no-restore --quiet CMD check  \
  '/var/folders/wn/rm6ws4bn3xx_ll9b5x3gy9nr0000gn/T//RtmpNrTKVp/rete_0.1.0.tar.gz'  \
  --as-cran --timings --no-manual 

* using log directory ‘/Users/farzantaj/Second Semester/BCB420/rete.Rcheck’
* using R version 3.3.3 (2017-03-06)
* using platform: x86_64-apple-darwin13.4.0 (64-bit)
* using session charset: UTF-8
* using options ‘--no-manual --as-cran’
* checking for file ‘rete/DESCRIPTION’ ... OK
* checking extension type ... Package
* this is package ‘rete’ version ‘0.1.0’
* package encoding: UTF-8
* checking package namespace information ... OK
* checking package dependencies ... OK
* checking if this is a source package ... OK
* checking if there is a namespace ... OK
* checking for executable files ... OK
* checking for hidden files and directories ... OK
* checking for portable file names ... OK
* checking for sufficient/correct file permissions ... OK
* checking whether package ‘rete’ can be installed ... OK
* checking installed package size ... OK
* checking package directory ... OK
* checking DESCRIPTION meta-information ... OK
* checking top-level files ... OK
* checking for left-over files ... OK
* checking index information ... OK
* checking package subdirectories ... OK
* checking R files for non-ASCII characters ... OK
* checking R files for syntax errors ... OK
* checking whether the package can be loaded ... OK
* checking whether the package can be loaded with stated dependencies ... OK
* checking whether the package can be unloaded cleanly ... OK
* checking whether the namespace can be loaded with stated dependencies ... OK
* checking whether the namespace can be unloaded cleanly ... OK
* checking loading without being on the library search path ... OK
* checking dependencies in R code ... OK
* checking S3 generic/method consistency ... OK
* checking replacement functions ... OK
* checking foreign function calls ... OK
* checking R code for possible problems ... OK
* checking Rd files ... OK
* checking Rd metadata ... OK
* checking Rd line widths ... OK
* checking Rd cross-references ... OK
* checking for missing documentation entries ... OK
* checking for code/documentation mismatches ... OK
* checking Rd \usage sections ... OK
* checking Rd contents ... OK
* checking for unstated dependencies in examples ... OK
* checking examples ... OK
* checking for unstated dependencies in ‘tests’ ... OK
* checking tests ...
  Running ‘testthat.R’ OK
* DONE
Status: OK

 [110s/112s]


R CMD check results
0 errors | 0 warnings | 0 notes

R CMD check succeeded
```

**Comment by hyginn** [comment]
Won't pursue.

**Pull request closed without merging**